sync/class.jetpack-sync-actions.php

```diff
@@ -440,26 +440,33 @@ static function cleanup_on_upgrade( $new_version = null, $old_version = null ) {
 	}
 
 	static function get_sync_status() {
+		require_once JETPACK__PLUGIN_DIR . 'sync/class.jetpack-sync-wp-replicastore.php';
 		self::initialize_sender();
 
 		$sync_module     = Jetpack_Sync_Modules::get_module( 'full-sync' );
 		$queue           = self::$sender->get_sync_queue();
 		$full_queue      = self::$sender->get_full_sync_queue();
 		$cron_timestamps = array_keys( _get_cron_array() );
 		$next_cron       = $cron_timestamps[0] - time();
+		$store           = new Jetpack_Sync_WP_Replicastore();
+		$checksums       = $store->checksum_all();
```

Comment:
How expensive is this method? Is there a situation where we would call this endpoint but not care about the checksums?

If the answers are: "pretty expensive: it's slow" and "yes, we sometimes don't care about the checksums", then we should write this in such a way that if `?fields=` is set and does not include any of the checksum fields, we don't bother calculating the checksums.

If it's cheap or if we know we always want this information, it doesn't matter.

Reply:
Good call. AFAIK, the SQL queries may be rather expensive as they calculate the checksum for all posts, comments and post_meta records.

However, calling this endpoint is slow anyway, but I could try timing it before and after this change.

I'm going to follow-up on this within Poseidon to coordinate on a way forward here. Using `?fields=` is possibly a good idea.

Reply:
[I did the research and crunched the numbers](https://docs.google.com/spreadsheets/d/10R94tmQZEHaDBgrE-vwqf0ybxCFIU7M_JFWhhp6kVmo/edit#gid=0). Note that this is made on my recent MBP development machine, using our Jetpack Docker image.

The relative overhead created by `Jetpack_Sync_WP_Replicastore::checksum_all()`, while high for smaller instances, becomes lower and lower the higher the number of posts is.

Initially, up to 200 posts (with increments of ca 30), the overhead added ranges between 30%-45%.

However, when we get to the tens of thousands, with increments of 5000-1000, the overhead becomes closer to 5%-15%.

The absolute range of this overhead in milliseconds ranged from 0.0032 ms to 0.0070 ms, so I really don't think this adds any significant overhead.

Looks at this fancy graph!

![Delta  %  vs Number of posts(1)](https://user-images.githubusercontent.com/191583/57084679-09e36780-6cfb-11e9-8d31-cec0b0ef528d.png)

Reply:
Note that on the WPCOM side, I could have done a similar timing test, but I think my internet connection in combination with tunneling services that we use such as Pagekite and ngrok add too much of a response lag to make be able to trust such tests.

Reply:
@aldavigdis
I don’t understand how the as the numbers of posts increases the call with the non checksums also gets slower. 
Do you have an explanation for that? 
Is it also possible to remove any dependence on the network in this performance tests. Since that can have a lot of variance in there as well.

Reply:
This was all done within the `Jetpack_Sync_Actions::get_sync_status` function, the diff is here:

```diff
diff --git a/sync/class.jetpack-sync-actions.php b/sync/class.jetpack-sync-actions.php
index 79c47b9c1..7f229342a 100644
--- a/sync/class.jetpack-sync-actions.php
+++ b/sync/class.jetpack-sync-actions.php
@@ -440,6 +440,7 @@ class Jetpack_Sync_Actions {
        }
 
        static function get_sync_status() {
+               $time_start = microtime( true );
                require_once JETPACK__PLUGIN_DIR . 'sync/class.jetpack-sync-wp-replicastore.php';
                self::initialize_sender();
 
@@ -451,7 +452,13 @@ class Jetpack_Sync_Actions {
                $store           = new Jetpack_Sync_WP_Replicastore();
                $checksums       = $store->checksum_all();
 
+               $time_taken_before_sync = microtime( true ) - $time_start;
+
                $full_sync_status = ( $sync_module ) ? $sync_module->get_status() : array();
+
+               $time_taken = microtime( true ) - $time_start;
+               l( wp_count_posts()->publish . ' ' . $time_taken . ' ' . $time_taken_before_sync );
+
                return array_merge(
                        $full_sync_status,
                        array(

```

If there is a better way to do this without too much hassle, then I'm all ears.

Reply:
For the record, I have done this as of ac1d2f8ac68b6fa705501f5316dcfc462f6443f8.

The checksums are not included and not crunched by the helper method unless those keys are specifically requested using the `filter` parameter.

Example:

```
/sites/site_id/sync/status?fields=posts_checksum,comments_checksum,post_meta_checksum,comment_meta_checksum
```